[16.0][IMP] sale_order_general_discount_triple

RabbitJon-S73 opened the pull request:
Pricelist discount is override by global discount, With this improve pricelist discount can be applied in any discount like global discount does

RabbitJon-S73 commented:
@rousseldenis could you review this contribution?

RabbitJon-S73 commented:
@AshishHirapara could you review? Thanks!

rousseldenis commented on sale_order_general_discount_triple/models/sale_order_line.py:
```diff
@@ -52,3 +52,31 @@ def create(self, vals):
         return super().create(vals)
+
+    @api.onchange("product_id", "product_uom_qty", "product_uom")
+    def _onchange_product_id_discount(self):
```
@RabbitJon-S73 I would avoid onchanges by changing fields properties to compute, stored=True, readonly=False

rousseldenis commented on sale_order_general_discount_triple/tests/test_module.py:
```diff
@@ -2,26 +2,42 @@
-        partner_form.name = "Test Partner"
-        partner = partner_form.save()
-        setting_form = Form(setting_obj)
+    def setUp(self):
```
Change this to classmethod

RabbitJon-S73 commented:
@rousseldenis changed

rousseldenis commented on sale_order_general_discount_triple/__manifest__.py:
```diff
@@ -3,7 +3,7 @@
     "name": "Sale Order General Discount Triple",
     "summary": "General discount per sale order with triple",
-    "version": "16.0.1.0.0",
+    "version": "16.0.1.1.0",
```
Unless you have migration scripts that depends on this, this is not necessary, done by ocabot

RabbitJon-S73 replied:
I didn't know. Its done

RabbitJon-S73 commented:
@rousseldenis if everything its okey now could we merge this PR?

hildickethan commented on sale_order_general_discount_triple/models/sale_order_line.py:
```diff
@@ -4,51 +4,81 @@
-    def default_get(self, fields):
-        vals = super().default_get(fields)
-        general_discount = (
+    discount2 = fields.Float(compute="_compute_discount2", store=True, readonly=False)
+    discount3 = fields.Float(compute="_compute_discount3", store=True, readonly=False)
```
these need to be `precompute=True` to avoid these warnings
 `Field sale.order.line.price_reduce_taxinc cannot be precomputed as it depends on non-precomputed field sale.order.line.price_total`
`Field sale.order.line.price_total cannot be precomputed as it depends on non-precomputed field sale.order.line.discount2`
and a few more caused by the same root
I will push the change

RabbitJon-S73 commented:
@rousseldenis I splited the changes on this PR in two commits, one for fixing the issue commented on https://github.com/OCA/sale-workflow/pull/2358 and the other one for the improvements.


@pedrobaeza

pedrobaeza commented on sale_order_general_discount_triple/models/sale_order_line.py:
```diff
@@ -4,51 +4,44 @@
-        )
+    @api.depends("product_id", "product_uom", "product_uom_qty")
+    def _compute_discount(self):
+        if hasattr(super(), "_compute_discount"):
```
I don't think you need this if, as the compute is on the module that declares the field, isn't it?

pedrobaeza commented on sale_order_general_discount_triple/models/sale_order_line.py:
```diff
@@ -4,51 +4,44 @@
+        general_discount = self._get_general_discount_field()
+        for line in self:
+            vals = {"discount3": 0.0}
+            if general_discount == "discount3":
```
Exit early if the discount field is not the one we are looking for, and you can pre-assign `self.discount3 = 0`

RabbitJon-S73 replied:
Done!

RabbitJon-S73 committed:
[FIX] sale_order_general_discount_triple: make all discounts computed

Discount behaviuour in v16.0 is to be computed but in this module it was changed. Here we are using compute_discount method to archive the functionallity and also converting the remaining two discounts to computed fields.

RabbitJon-S73 committed:
[IMP] sale_order_general_discount_triple: Add pricelist discount

Pricelist discount was overide by general discount functionality. With this change, its up to the user to choose in which discount wants to apply general and pricelist disounts.

RabbitJon-S73 commented:
@pedrobaeza @rousseldenis PR ready to merge

hildickethan commented:
Since this fixes the compute issues you can change the `installable` in the manifest back to `True`
And also would be cleaner to cut off the `<` typo from this commit 
![image](https://github.com/OCA/sale-workflow/assets/47854752/5a5e21fb-e8d1-4109-be34-959169b02569)

francesco-ooops commented on sale_order_general_discount_triple/README.rst:
```diff
@@ -38,8 +38,10 @@ Usage
 #. Go to Settings > General Settings > Pricing
-#. Under pricing section, select your prefered Discount type from 'General Discount' field.
-#. The selected discount type among the 3 discounts will be automatically parsed on sale order line while using 'Discount %' text field.
+#. Under pricing section, select your prefered Discount type from 'General Discount' field and
+   the same for 'Pricelist Discount'.
+#. The selected global discount type among the 3 discounts will be automatically parsed on sale order line while using 'Discount %' text field.
+#. If pricelist has a discount, it will be applied on the selected pricelist discount.
```
@RabbitJon-S73 can you add this change in the USAGE file?

RabbitJon-S73 replied:
Done! I also add some info that can be usefull

francesco-ooops commented:
@RabbitJon-S73 question: what is the purpose/behavior of the option "no apply" for "pricelist discount" setting? I think it can be clarified in usage and/or description

francesco-ooops commented:
@RabbitJon-S73 on a side note, you may want to take a look at this behavior: 

- set "general discount" in "discount 2" (settings)
- in SO, add a product and a general discount
- now add shipping expenses to SO

CURRENT BEHAVIOR

Discount for shipping product is set on Discount 1 and not Discount 2

![image](https://github.com/OCA/sale-workflow/assets/55489873/e947118f-f99d-4ac4-82f3-1b4871181c35)

FYI we're trying to fix this in #2666

RabbitJon-S73 commented:
> @RabbitJon-S73 question: what is the purpose/behavior of the option "no apply" for "pricelist discount" setting? I think it can be clarified in usage and/or description

I did that in order to let the user choose the "old behaviour" with no pricelist discounts. What do you think about doing the same with general discount to allow disable it?

RabbitJon-S73 commented:
> @RabbitJon-S73 on a side note, you may want to take a look at this behavior:
> 
>     * set "general discount" in "discount 2" (settings)
> 
>     * in SO, add a product and a general discount
> 
>     * now add shipping expenses to SO
> 
> 
> CURRENT BEHAVIOR
> 
> Discount for shipping product is set on Discount 1 and not Discount 2
> 
> ![image](https://user-images.githubusercontent.com/55489873/264964162-e947118f-f99d-4ac4-82f3-1b4871181c35.png)
> 
> FYI we're trying to fix this in #2666

I see your example in runboat but i can't reproduce it anymore, could you give me more details? The only I noticed is that general discount is not applied on discount2 of delivery line because the compute is not triggered. I'm trying to fix it

francesco-ooops commented:
> > @RabbitJon-S73 question: what is the purpose/behavior of the option "no apply" for "pricelist discount" setting? I think it can be clarified in usage and/or description
> 
> I did that in order to let the user choose the "old behaviour" with no pricelist discounts. What do you think about doing the same with general discount to allow disable it?

Can you explain what you mean by "old behaviour with no pricelist discounts"?

francesco-ooops commented:
> I see your example in runboat but i can't reproduce it anymore, could you give me more details? The only I noticed is that general discount is not applied on discount2 of delivery line because the compute is not triggered. I'm trying to fix it

You can check this video: https://recordit.co/R6S2YKT2Id (you can see there are also other behaviors that don't seem right...)

RabbitJon-S73 commented:
> > > @RabbitJon-S73 question: what is the purpose/behavior of the option "no apply" for "pricelist discount" setting? I think it can be clarified in usage and/or description
> > 
> > 
> > I did that in order to let the user choose the "old behaviour" with no pricelist discounts. What do you think about doing the same with general discount to allow disable it?
> 
> Can you explain what you mean by "old behaviour with no pricelist discounts"?

Before this PR, general discount overwrite the pricelist discount, i suppose that is not desired but I think that could be useful for some users to disable pricelist discount or general discount by choosing the option to "No apply"

RabbitJon-S73 commented:
> > I see your example in runboat but i can't reproduce it anymore, could you give me more details? The only I noticed is that general discount is not applied on discount2 of delivery line because the compute is not triggered. I'm trying to fix it
> 
> You can check this video: https://recordit.co/R6S2YKT2Id (you can see there are also other behaviors that don't seem right...)

I think I fixed that, but I'm not happy with the solution since I had to call compute functions from the method that creates the delivery line. If someone finds a better solution it's welcome

francesco-ooops commented:
> > > > @RabbitJon-S73 question: what is the purpose/behavior of the option "no apply" for "pricelist discount" setting? I think it can be clarified in usage and/or description
> > > 
> > > 
> > > I did that in order to let the user choose the "old behaviour" with no pricelist discounts. What do you think about doing the same with general discount to allow disable it?
> > 
> > 
> > Can you explain what you mean by "old behaviour with no pricelist discounts"?
> 
> Before this PR, general discount overwrite the pricelist discount, i suppose that is not desired but I think that could be useful for some users to disable pricelist discount or general discount by choosing the option to "No apply"

I don't think it's good ux to allow users to set discounts in pricelist and disabling them through a general setting. 

I think having a setting (disc 1, 2 or 3) for each type of discount (general and pricelist) is sufficient, but I would leave both fields in settings empty and required to force user to set them up correctly.

RabbitJon-S73 committed:
[IMP] sale_order_general_discount_triple: style

RabbitJon-S73 commented:
@francesco-ooops @pedrobaeza can be merged?

pedrobaeza commented:
/ocabot merge major

celm1990 commented:
@RabbitJon-S73  delete default value but fields are `required` cause issue on other modules, please can review other alternative(remove attribute `required`?)
https://github.com/OCA/sale-workflow/pull/2392/files#diff-06b9aec82dbbd6bf4122a463642bbeac0e486e2e760e5cf336b3a18c153fe43aL14

See https://github.com/OCA/sale-workflow/actions/runs/6808350353/job/18512649384
![image](https://github.com/OCA/sale-workflow/assets/7775116/cab87470-1ffb-440b-bf32-f8d91ae797dc)
